File: ktorOld/main/kotlin/schema/mutation/StorageMutation.kt
package schema.mutation

import com.expediagroup.graphql.generator.annotations.GraphQLDescription
import com.expediagroup.graphql.server.operations.Mutation
import services.StorageService
import services.StorageDbService
import schema.model.*
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

@Suppress("unused")
class StorageMutation : Mutation, KoinComponent {
    private val storageService: StorageService by inject()
    private val storageDbService: StorageDbService by inject()

    @GraphQLDescription("Rename a file")
    fun renameFile(
        input: RenameFileInput
    ): FileOperationResult {
        return storageService.renameFile(input)
    }

    @GraphQLDescription("Delete a single file")
    fun deleteFile(
        path: String
    ): FileOperationResult {
        return storageService.deleteFile(path)
    }

    @GraphQLDescription("Generate a signed URL for file upload")
    fun generateUploadSignedUrl(
        input: GenerateUploadSignedUrlInput
    ): String {
        val path = "${input.location.path}/${input.fileName}"
        return storageService.generateUploadSignedUrl(path, input.contentType)
    }

    @GraphQLDescription("Create a new folder")
    fun createFolder(
        input: CreateFolderInput
    ): FileOperationResult {
        return storageService.createFolder(input)
    }

    @GraphQLDescription("Move multiple files/folders to a new location")
    suspend fun moveStorageItems(
        input: MoveStorageItemsInput
    ): BulkOperationResult {
        return storageService.moveItems(input)
    }

    @GraphQLDescription("Copy multiple files/folders to a new location")
    suspend fun copyStorageItems(
        input: CopyStorageItemsInput
    ): BulkOperationResult {
        return storageService.copyItems(input)
    }

    @GraphQLDescription("Delete multiple files/folders")
    suspend fun deleteStorageItems(
        input: DeleteItemsInput
    ): BulkOperationResult {
        return storageService.deleteItems(input)
    }

    @GraphQLDescription("Update directory permissions")
    suspend fun updateDirectoryPermissions(
        input: UpdateDirectoryPermissionsInput
    ): FileOperationResult {
        return storageDbService.updateDirectoryPermissions(input)
    }

    @GraphQLDescription("Set protection for files or directories")
    suspend fun setStorageItemProtection(
        input: SetStorageItemProtectionInput
    ): FileOperationResult {
        return storageDbService.setProtection(input)
    }
}

File: server/graphql/storage/storage.mutation.ts
import { gqlSchemaBuilder } from "../gqlSchemaBuilder";
import * as StoragePothos from "./storage.pothos";
import { getStorageService } from "../../storage/storage.service";
import { StorageDbRepository } from "../../storage/db/storage-db.service";

gqlSchemaBuilder.mutationFields((t) => ({
    renameFile: t.field({
        type: StoragePothos.FileOperationResultPothosObject,
        args: {
            input: t.arg({
                type: StoragePothos.FileRenameInputPothosObject,
                required: true,
            }),
        },
        resolve: async (_parent, { input }) => {
            const storageService = await getStorageService();
            return await storageService.renameFile(input);
        },
    }),

    deleteFile: t.field({
        type: StoragePothos.FileOperationResultPothosObject,
        args: {
            path: t.arg.string({ required: true }),
        },
        resolve: async (_parent, { path }) => {
            const storageService = await getStorageService();
            return await storageService.deleteFile(path);
        },
    }),

    generateUploadSignedUrl: t.field({
        type: "String",
        args: {
            input: t.arg({
                type: StoragePothos.GenerateUploadSignedUrlInputPothosObject,
                required: true,
            }),
        },
        resolve: async (_parent, { input }) => {
            const storageService = await getStorageService();
            return await storageService.generateUploadSignedUrl(input);
        },
    }),

    createFolder: t.field({
        type: StoragePothos.FileOperationResultPothosObject,
        args: {
            input: t.arg({
                type: StoragePothos.FolderCreateInputPothosObject,
                required: true,
            }),
        },
        resolve: async (_parent, { input }) => {
            const storageService = await getStorageService();
            return await storageService.createFolder(input);
        },
    }),

    moveStorageItems: t.field({
        type: StoragePothos.BulkOperationResultPothosObject,
        args: {
            input: t.arg({
                type: StoragePothos.StorageItemsMoveInputPothosObject,
                required: true,
            }),
        },
        resolve: async (_parent, { input }) => {
            const storageService = await getStorageService();
            return await storageService.moveItems(input);
        },
    }),

    copyStorageItems: t.field({
        type: StoragePothos.BulkOperationResultPothosObject,
        args: {
            input: t.arg({
                type: StoragePothos.StorageItemsCopyInputPothosObject,
                required: true,
            }),
        },
        resolve: async (_parent, { input }) => {
            const storageService = await getStorageService();
            return await storageService.copyItems(input);
        },
    }),

    deleteStorageItems: t.field({
        type: StoragePothos.BulkOperationResultPothosObject,
        args: {
            input: t.arg({
                type: StoragePothos.StorageItemsDeleteInputPothosObject,
                required: true,
            }),
        },
        resolve: async (_parent, { input }) => {
            const storageService = await getStorageService();
            return await storageService.deleteItems(input);
        },
    }),

    updateDirectoryPermissions: t.field({
        type: StoragePothos.FileOperationResultPothosObject,
        args: {
            input: t.arg({
                type: StoragePothos.DirectoryPermissionsUpdateInputPothosObject,
                required: true,
            }),
        },
        resolve: async (_parent, { input }) => {
            return await StorageDbRepository.updateDirectoryPermissions(input);
        },
    }),

    setStorageItemProtection: t.field({
        type: StoragePothos.FileOperationResultPothosObject,
        args: {
            input: t.arg({
                type: StoragePothos.StorageItemProtectionUpdateInputPothosObject,
                required: true,
            }),
        },
        resolve: async (_parent, { input }) => {
            return await StorageDbRepository.setProtection(input);
        },
    }),
}));
